GraphResponse.md

```diff
@@ -7,6 +7,206 @@ Provide a decorator for the native response message to enable attaching Graph sp
+
+These would in turn call a function in the core library(BaseRequest) that looks something like
+
+```csharp
+
+    public async Task<GraphResponse> SendAsyncWithGraphResponse() {
+        using (var response = await this.SendRequestAsync(serializableObject, cancellationToken, completionOption).ConfigureAwait(false))
+        {
+            return new GraphResponse(this,response);
+        }
+    }
+
+```
+### Example Usage
+
+Example calls will look like this.
+
+```csharp
+
+GraphResponse response = await graphServiceClient.Users.Request().GetWithGraphResponseAsync(cancellationToken);
+
+GraphResponse response = await graphServiceClient.Me.Request().UpdateWithGraphResponseAsync(patchUser, cancellationToken);
+
+```
+
+### Other Notes
+
+* Delete requests can (by default) return GraphResponse objects rather than void
+* Pro: Use generator to get capabilty annotation generation support.
+* Cons: T4 template code to maintain.
+* We can possibly add extension methods to GraphResponse to also return domain objects
+
+## Late and Custom Deserialization
+
+This can be done exposing the `IResponseHandler` interface as below. (This will involve some tweaking of the current ResponseHandlers as well).
+
+```csharp
+
+    /// <summary>
+    /// The interface required for all response handlers.
+    /// </summary>
+    public interface IResponseHandler
+    {
+        /// <summary>
+        /// Process raw HTTP response into the requested domain type.
+        /// </summary>
+        /// <typeparam name="T">The type to return</typeparam>
+        /// <param name="response">The HttpResponseMessage to handle.</param>
+        /// <returns></returns>
+        Task<T> HandleResponse<T>(HttpResponseMessage response);
+
+        /// <summary>
+        /// Process the GraphResponse into the requested domain type.
+        /// </summary>
+        /// <typeparam name="T">The type to return</typeparam>
+        /// <param name="graphResponse">The GraphResponse to handle.</param>
+        /// <returns></returns>
+        Task<T> HandleResponse<T>(GraphResponse graphResponse);
+    }
+
+```
+
+This means that customers could use a custom serializer while implementing the interface and simply throw in a `GraphResponse` object into it. 
+If we align our current ResponseHandler to this, they could possibly just use our ResponseHandler as well.
+
+Therefore a developer could possibly follow the following steps.
+
+```csharp
+
+    ISerializer serailizer = new CustomSerializer(); //Custom Serializer
+    IResponseHandler responseHandler = new ResponseHandler(serailizer); // Our ResponseHadler with custom Serializer
+    
+    GraphResponse graphResponse = await graphServiceClient.Me.Request().UpdateWithGraphResponseAsync(patchUser, cancellationToken);//response with no serialization
+    
+    //
+    // Do other stuff here like check status/headers
+    //
+
+    User me = responseHandler.HandleResponse<User>(graphResponse); // Late and custom deserialization
```

Comment:
Problem here is that the customer has to magically know the type parameter. `UpdateWithGraphResponseAsync` can be generated with the type parameter which can be passed to the GraphResponse.

`GraphResponse<User> graphResponse = await graphServiceClient.Me.Request().UpdateWithGraphResponseAsync<User>(patchUser, cancellationToken);`

Reply:
This makes sense to me. I have revised this appropiately.